sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -53,6 +58,10 @@ private DoFnSignatures() {}
 
   private final Map<Class<?>, DoFnSignature> signatureCache = new LinkedHashMap<>();
 
+  public synchronized <FnT extends DoFn<?, ?>> DoFnSignature signatureForDoFn(FnT fn) {
```

Comment:
`/** @return the {@link DoFnSignature} for the given {@link DoFn} instance. */`

Also, doesn't need to be synchronized -- calling `fn.getClass()` is thread safe, and `getOrParseSignature` is synchronized.